internal/controller/managedzone_controller.go

```diff
@@ -186,21 +204,21 @@ func (r *ManagedZoneReconciler) deleteManagedZone(ctx context.Context, managedZo
 
 	dnsProvider, err := r.ProviderFactory.ProviderFor(ctx, managedZone, provider.Config{})
 	if err != nil {
-		var reason, message string
-		status := metav1.ConditionFalse
-		reason = "ProviderError"
-		message = fmt.Sprintf("The DNS provider creation failed: %v", err)
-		managedZone.Status.ObservedGeneration = managedZone.Generation
-		setManagedZoneCondition(managedZone, string(v1alpha1.ConditionTypeReady), status, reason, message)
-		return err
+		return ZoneErr{
+			code:    ZoneErrInternal,
+			message: fmt.Sprintf("failed to get DNS provider instance : %v", err),
+		}
 	}
 	err = dnsProvider.DeleteManagedZone(managedZone)
 	if err != nil {
 		if strings.Contains(err.Error(), "was not found") || strings.Contains(err.Error(), "notFound") {
 			log.Log.Info("ManagedZone was not found, continuing", "managedZone", managedZone.Name)
 			return nil
 		}
-		return err
+		return ZoneErr{
+			code:    ZoneErrProvider,
+			message: fmt.Sprintf("Failed to delete from provider. Provider Error: %v", provider.SanitizeError(err)),
+		}
```

Comment:
Is this preferable to just wrapping an error with a known type?

```.go
var (
   ErrProvider = errors.New("ProviderError")
)
...
if err != nil {
   return fmt.Errorf("%w, failed to get provider for the zone: %v", ErrProvider, provider.SanitizeError(err))
}
```

and using errors.Is where you want to check the type:

```go
if errors.Is(ErrProvider, err) {
   reason = ErrProvider.Error()
   message = err.Error()
}
```